SimTKcommon/src/Pathname.cpp

```diff
@@ -110,94 +110,117 @@ static void removeDriveInPlace(string& inout, string& drive) {
+        processed = pathcleaned;
+        if (!pathdrive.empty()) swddrive = "";
+    }
+    else if (pathcleaned.substr(0, 2) == "./") {
+        if (!swd.empty()) {
+            pathcleaned.erase(0, 2);
+            processed = swdcleaned + pathcleaned;
+        }
+        else {
+            processed = pathcleaned;
+        }
+    }
+    // Looks like a relative path name.
+    else {
+        processed = swdcleaned + pathcleaned;
+    }
 
     // If the pathname in its entirety is just one of these, append 
     // a slash to avoid special cases below.
-    if (processed=="." || processed==".." || processed=="@")
+    if (processed == "." || processed == ".." || processed == "@")
         processed += "/";
 
     // If the path begins with "../" we'll make it ./../ to simplify handling.
-    if (processed.substr(0,3) == "../")
+    if (processed.substr(0, 3) == "../")
         processed.insert(0, "./");
 
-    if (processed.substr(0,1) == "/") {
-        isAbsolutePath = true;
-        processed.erase(0,1);
-        if (drive.empty()) drive = getCurrentDriveLetter();
-    } else if (processed.substr(0,2) == "./") {
-        isAbsolutePath = true;
-        processed.replace(0,2,getCurrentWorkingDirectory(drive));
-        removeDriveInPlace(processed, drive);
-    } else if (processed.substr(0,2) == "@/") {
-        isAbsolutePath = true;
-        processed.replace(0,2,getThisExecutableDirectory());
-        removeDriveInPlace(processed, drive);
-    } else if (!drive.empty()) {
-        // Looks like a relative path name. But if it had an initial
-        // drive specification, e.g. X:something.txt, that is supposed
-        // to be interpreted relative to the current working directory
-        // on drive X, just as though it were X:./something.txt.
-        isAbsolutePath = true;
-        processed.insert(0, getCurrentWorkingDirectory(drive));
-        removeDriveInPlace(processed, drive);
+    if (processed.substr(0, 1) == "/") {
+        dontApplySearchPath = true;
+        processed.erase(0, 1);
+        if (!swddrive.empty()) finaldrive = swddrive;
+        else if (!pathdrive.empty()) finaldrive = pathdrive;
+        else finaldrive = getCurrentDriveLetter();
+    }
+    else if (processed.substr(0, 2) == "./") {
+        dontApplySearchPath = true;
+        String discard;
+        // If swd was empty, then "./" came from path. Make sure not to write
+        // over current pathdrive or swddrive.
+        if (swd.empty()) {
+            processed.replace(0, 2, getCurrentWorkingDirectory(pathdrive));
+            removeDriveInPlace(processed, discard);
+        }
+        // Otherwise, "./" came from swd
+        else {
+            processed.replace(0, 2, getCurrentWorkingDirectory(swddrive));
+            removeDriveInPlace(processed, discard);
+        }
+        if (!swddrive.empty()) finaldrive = swddrive;
+        else if (!pathdrive.empty()) finaldrive = pathdrive;
+        else finaldrive = getCurrentDriveLetter();
```

Comment:
There is something wrong with the logic here. Why is it so complicated compared to the "@/" case?